web/src/main/webapp/ui/menu.ts

```diff
@@ -130,6 +160,56 @@ abstract class BaseMenu<MI extends BaseMenuItem> implements IHtmlElement {
         return cell;
     }
 
+    /**
+     * Inserting subMenu must be done in the end of all parent menu insert. If the
+     * submenu exceed the length of all main menu, we need to add dummyMenu
+     * */
+    public insertSubMenu(parentText: string, mi: MI, enabled: boolean): void {
+        const parentIndex = this.find(parentText);
+        if (parentIndex < 0) throw new Error("Cannot find menu item " + parentText);
+        if (this.subMenuCells[parentIndex] === undefined)
+            this.subMenuCells[parentIndex] = emptySubMenu();
+
+        const subMenuIndex = this.subMenuCells[parentIndex].cells.length;
+        const subMenuPlacementIdx = parentIndex + subMenuIndex;
+        if (this.rows[subMenuPlacementIdx] === undefined) {
+            const trow = this.tableBody.insertRow();
+            this.rows.push(trow);
+            const cell = trow.insertCell(0);
+            cell.classList.add("dummyMenu");
+        }
+        const cell = this.rows[subMenuPlacementIdx].insertCell(1);
+        cell.id = makeId(mi.text);
+        cell.style.textAlign = "left";
+        cell.style.display = "none";
+        cell.classList.add("menuItem");
+        if (enabled) cell.classList.remove("disabled");
+        else cell.classList.add("disabled");
+        if (mi.help != null) cell.title = mi.help;
+        if (mi.text === "---") cell.innerHTML = "<hr>";
+        else cell.textContent = mi.text;
+
+        var subIndex = this.subMenuCells[parentIndex].cells.length;
+        this.subMenuCells[parentIndex].cells.push(cell);
+        cell.onmouseenter = () => this.selectSubMenu(parentIndex, subIndex);
+        cell.onmouseleave = () => this.selectSubMenu(parentIndex, -1);
+        this.setSubMenuAction(parentIndex, subMenuIndex, mi, enabled);
+    }
+
+    public addExpandableItem(mi: MI, enabled: boolean): HTMLTableDataCellElement {
```

Comment:
I noticed that when the popup menu is displayed the menu items below all shift down. So there is something wrong with the styling.

Reply:
Yes, because of the border and shadow that add more height

Reply:
Addressed this by changing the display from `block` to `table-cell`